app/src/examples/KeyboardAvoidingViewExample.tsx

```diff
@@ -0,0 +1,251 @@
+      state.value,
+      'modalVisible.value',
+      modalVisible.value
+    );
+
+    if (state.value === KeyboardState.OPEN && !modalVisible.value) {
+      ctx.lastHeight = height.value;
+    }
+
+    if (state.value === KeyboardState.CLOSING) {
+      ctx.lastHeight = 0;
+    }
+
+    if (modalVisible.value || state.value === KeyboardState.CLOSED) {
+      return Math.max(ctx.lastHeight - 17, 0);
+    }
+
+    return 0;
+  });
+
+  const animatedStyle = useAnimatedStyle(() => {
+    return {
+      transform: [
+        {
+          translateY: translateY.value,
+        },
+      ],
+    };
+  });
+
+  return (
+    <ScrollView {...props}>
+      <Animated.View style={[{ flex: 1 }, animatedStyle]}>
+        {children}
+      </Animated.View>
+    </ScrollView>
+  );
+}
+
+export default function KeyboardAvoidingViewExample(): React.ReactElement {
+  const insets = useSafeAreaInsets();
+
+  const [useReanimated, setUseReanimated] = useState(false);
+
+  const [isModalVisible, setIsModalVisible] = useState(false);
+  const modalVisible = useSharedValue(false);
+
+  const showModal = useCallback(() => {
+    runOnUI(() => {
+      'worklet';
+
+      modalVisible.value = true;
+    })();
+
+    setIsModalVisible(true);
+  }, [modalVisible, setIsModalVisible]);
+
+  const hideModal = useCallback(() => {
+    runOnUI(() => {
+      'worklet';
+
+      modalVisible.value = false;
+    })();
+
+    setIsModalVisible(false);
+  }, [modalVisible, setIsModalVisible]);
+
+  const renderScrollComponent = useCallback(
+    (props: ScrollViewProps) => {
+      // @ts-expect-error
+      return <KeyboardSpace {...props} modalVisible={modalVisible} />;
+    },
+    [modalVisible]
+  );
+
+  const KeyboardAvoidingView = useReanimated
+    ? // @ts-expect-error
+      Animated.KeyboardAvoidingView
+    : RNKeyboardAvoidingView;
+
+  return (
+    <View style={[styles.container]}>
+      <View style={styles.header}>
+        <Text>Use Reanimated KeyboardAvoidingView</Text>
+        <Switch
+          value={useReanimated}
+          onChange={(evt) => setUseReanimated(evt.nativeEvent.value)}
+        />
+      </View>
+
+      <FlatList
+        inverted
+        renderScrollComponent={renderScrollComponent}
+        keyboardDismissMode="on-drag"
+        keyboardShouldPersistTaps="always"
+        data={Array.from({ length: 20 }).map((_, i) => i)}
+        renderItem={({ item }) => (
+          <ListItem onLongPress={showModal} title={`Item ${item}`} />
+        )}
+      />
+
+      <KeyboardAvoidingView
+        behavior="padding"
+        keyboardVerticalOffset={40 + insets.bottom}>
+        <View
+          style={{
+            paddingBottom: insets.bottom,
+          }}>
+          <TextInput
+            placeholder="Message here..."
+            style={[styles.textInput]}
+            autoCorrect
+          />
+        </View>
+      </KeyboardAvoidingView>
+
+      <Modal
+        style={styles.sheetContainer}
+        visible={isModalVisible}
```

Comment:
Hey @terrysahaidak, I've tested your example with the PR #4360.

If you look closely, there's still some flickering when opening the modal and it happens both with Reanimated's and React Native's KeyboardAvoidingView:

<details>
<summary>KeyboardAvoidingView from Reanimated</summary>

https://user-images.githubusercontent.com/20516055/232049370-4d3a7924-f0b4-4a35-939f-6447a37eb287.mp4

</details>

<details>
<summary>KeyboardAvoidingView from React Native</summary>

https://user-images.githubusercontent.com/20516055/232049437-dc6a5ce5-679e-4dbf-a1ba-47fafc093900.mp4

</details>

I believe the root cause of this flicker is not the implementation of `useAnimatedKeyboard` hook but rather the fact that the visibility of `<Modal>` component is controlled using React state whereas the other part of the logic uses a shared value. In particular, there are two sources of truth and it takes 1-2 frames to render the changes from the JS thread when the React state is updated:

https://github.com/software-mansion/react-native-reanimated/blob/d01993052c143ea95c321804277b35a3689f0172/app/src/examples/KeyboardAvoidingViewExample.tsx#L109-L110

Instead, you should keep the state only in shared value and control the visibility of the modal via animated props on the UI thread:

```tsx
import Animated, { useAnimatedProps } from 'react-native-reanimated';

const AnimatedModal = Animated.createAnimatedComponent(Modal);

const animatedProps = useAnimatedProps(() => {
  return { visible: isModalVisible.value };
});

return <AnimatedModal {/* props */} animatedProps={animatedProps} />;
```